hw/ip/pwm/doc/_index.md

```diff
@@ -0,0 +1,313 @@
+
+As a variant of the blink feature, the output duty cycle can also be programmed to linearly increase and decrease in time.
+In this "heartbeat" mode, the duty cycle increments by a programmable amount after a programmable number of pulses, starting at some firmware-selected initial duty cycle.
+Once the internal duty cycle reaches the target value, the internal duty cycle begins to decrement until it returns to the initial value, at which point the cycle repeats until heartbeat mode is disabled.
+
+# Theory of Operations
+
+## Block Diagram
+
+![](pwm_block_diagram.svg)
+
+## Hardware Interfaces
+
+{{< hwcfg "hw/ip/pwm/data/pwm.hjson" >}}
+
+## Design Details
+
+### Phase and Duty Cycle Representation
+
+The PWM IP creates series of pulses with the desired on-off duty cycle.
+The duty cycle, DC, is typically expressed a fraction of pulse duration, <i>d</i>, over the period between pulses, <i>T</i>:
+$$DC\equiv d/T$$.
+Since 0&lt;<i>d</i>&lt;<i>T</i>, the duty cycle generally ranges from 0 to 1.
+
+The PWM IP can control the duty cycle in a number of ways:
+- The PWM can be programmed to generate pulses at a firmware-defined duty cycle.
+- The duty cycle can be programmed to toggle (or "blink") between two duty cycles at a programmable rate.
+In the tri-color LED use-case, this would have the visual effect of making the LED blink between two colors.
+- The duty cycle can linearly sweep in time, gradually shifting back-and-forth between two endpoints.
+
+Thus the duty cycle can be controlled by firmware, or may change under hardware control.
+The operation of each of these modes will be discussed later.
+
+Each channel can also be assigned a different phase delay.
+Like the duty cycle, this delay is expressed as a fraction of the pulse period, T.
+The phase delay of each channel is always directly controlled by a firmware register value.
+
+Since the phase and duty cycle are always a fraction less than or equal to one, the PWM IP represents them as 16-bit fixed point numbers, with an implicit 16-bit shift.
+If the duty cycle is internally represented as a 16-bit value x, the output pulse train will have the duty cycle:
+$$DC(x)=\frac{x}{2^{16}}.$$
+Thus the allowed duty cycle in principle ranges from 0 to 1-1/2<sup>16</sup>.
+
+However, the actual phase resolution may be smaller.
+In order to support faster pulse rates, the phase resolution can be set to less than 16-bits, in which case the observed duty cycle will be rounded down to the next lowest multiple of 2<sup>-({{< regref "CFG.DC_RESN" >}}+1)</sup>.
+In other words, the {{< regref "CFG.DC_RESN" >}} register effectively limits the duty cycle resolution, such that only the {{< regref "CFG.DC_RESN" >}}+1 most significant bits are relevant:
+
+$$DC(x; \textrm{DC_RESN})=\frac{\textrm{MSB}(x, \textrm{DC_RESN}+1)}{2^{\textrm{DC_RESN}+1}},$$
+
+where here we use the notation MSB(<i>x</i>; <i>y</i>), to mean the <i>y</i> most significant bits of the binary value <i>x</i>.
```

Comment:
I think this should be `MSB(x, y)` (a comma, not a semicolon).

Reply:
Well, actually I did mean a semi-colon here as y is a function *parameter*, not an input.
So for consistency, I should probably make the other one a semi-colon..
However you are correct, they are inconsistent.  Thanks for catching that.